code/modules/reagents/chemistry/holder/holder.dm

```diff
@@ -333,6 +333,27 @@
 	handle_reactions()
 	return round(total_removed_amount, CHEMICAL_VOLUME_ROUNDING)
 
+/**
+ * Like remove_all but removes a percentage of every reagent directly rather than by volume
+ *
+ * Arguments
+ * * percentage - the percentage of each reagent to remove
+ */
+/datum/reagents/proc/remove_all_direct(percentage = 0.5)
+	if(!total_volume)
+		return 0
+	if(percentage <= 0)
+		stack_trace("non positive percentage passed to remove all reagents direct [percentage]")
+		return 0
+	var/list/cached_reagents = reagent_list
+	var/total_removed_amount = 0
+
+	for(var/datum/reagent/reagent as anything in cached_reagents)
```

Comment:
```suggestion
	var/total_removed_amount = 0

	for(var/datum/reagent/reagent as anything in reagent_list)
```
Frankly, it's already cached the moment you use it in such a loop iirc, and the file is already chock-full of this "cached_reagents" idiocy.